In [94]:
import xml.etree.ElementTree as ET

In [95]:
tree = ET.parse('Popular_baby_names.xml')
root = tree.getroot()

In [ ]:
for child in root.iter():
    print(child.tag)

In [96]:
rows=[]

for i in root.findall('row/row'):
  lst=[]
  id = i.attrib['_id']
  uuid = i.attrib['_uuid']
  address = i.attrib['_address']
   
  first_name = i.findtext('first_name')
  last_name = i.findtext('last_name')
  city = i.findtext('city')    
    
  lst.append(id)
  lst.append(uuid)
  lst.append(address)
  lst.append(first_name)
  lst.append(last_name)
  lst.append(city)
  rows.append(lst)
    
    
rows

[['row-ueyx_pbr6.ht3u',
  '00000000-0000-0000-7235-BFFC8FE5D830',
  'https://data.cityofnewyork.us/resource/mdcw-n682/row-ueyx_pbr6.ht3u',
  'Annette',
  'Charles',
  'New York'],
 ['row-5t8b_346w~6fsq',
  '00000000-0000-0000-33C9-8FD611F3B1E5',
  'https://data.cityofnewyork.us/resource/mdcw-n682/row-5t8b_346w~6fsq',
  'Keith',
  'Snow',
  'New York'],
 ['row-y52z~x7gt_5zsp',
  '00000000-0000-0000-0A9D-CB83FB2C9CFF',
  'https://data.cityofnewyork.us/resource/mdcw-n682/row-y52z~x7gt_5zsp',
  'Marek',
  'Tyszkiewicz',
  'New York'],
 ['row-e897~azvy.cjkb',
  '00000000-0000-0000-4E87-A63E672BFDA8',
  'https://data.cityofnewyork.us/resource/mdcw-n682/row-e897~azvy.cjkb',
  'Marlene',
  'Markoe-Boyd',
  'New York'],
 ['row-ie6x_vcbs_4rbs',
  '00000000-0000-0000-F1A9-90BD9EB88D53',
  'https://data.cityofnewyork.us/resource/mdcw-n682/row-ie6x_vcbs_4rbs',
  'Melissa',
  'Chacko',
  'New York'],
 ['row-usj8_gx7q-8hqd',
  '00000000-0000-0000-1BE4-BA4D8D63643D',
  'https://data.cityofnewyork.us/r

In [97]:
import pandas as pd
df= pd.DataFrame(rows ,columns=['ID', 'uuid', 'Address', 'First_name', 'Last_name', 'City'])
df

ID                                  uuid  \
0     row-ueyx_pbr6.ht3u  00000000-0000-0000-7235-BFFC8FE5D830   
1     row-5t8b_346w~6fsq  00000000-0000-0000-33C9-8FD611F3B1E5   
2     row-y52z~x7gt_5zsp  00000000-0000-0000-0A9D-CB83FB2C9CFF   
3     row-e897~azvy.cjkb  00000000-0000-0000-4E87-A63E672BFDA8   
4     row-ie6x_vcbs_4rbs  00000000-0000-0000-F1A9-90BD9EB88D53   
...                  ...                                   ...   
2531  row-w2ye~6ae8.ip8q  00000000-0000-0000-E302-06F2992D254C   
2532  row-nene.dygq.e72g  00000000-0000-0000-EE13-E543BEBD2BDD   
2533  row-8d7b-bf86_jrq9  00000000-0000-0000-0B1A-8CCF3CF56D98   
2534  row-nzmw_4y5g-htgt  00000000-0000-0000-B8D2-70472FD59060   
2535  row-q2yj~844r_3byi  00000000-0000-0000-724C-63423CED4A0E   

                                                Address First_name  \
0     https://data.cityofnewyork.us/resource/mdcw-n6...    Annette   
1     https://data.cityofnewyork.us/resource/mdcw-n6...      Keith   
2     https://data.cityofnewyork.us/resource/mdcw-n6...      Marek   
3     https://data.cityofnewyork.us/resource/mdcw-n6...    Marlene   
4     https://data.cityofnewyork.us/resource/mdcw-n6...    Melissa   
...                                                 ...        ...   
2531  https://data.cityofnewyork.us/resource/mdcw-n6...   Caroline   
2532  https://data.cityofnewyork.us/resource/mdcw-n6...     Denice   
2533  https://data.cityofnewyork.us/resource/mdcw-n6...     Darryl   
2534  https://data.cityofnewyork.us/resource/mdcw-n6...    Valerie   
2535  https://data.cityofnewyork.us/resource/mdcw-n6...      Susan   

        Last_name      City  
0         Charles  New York  
1            Snow  New York  
2     Tyszkiewicz  New York  
3     Markoe-Boyd  New York  
4          Chacko  New York  
...           ...       ...  
2531        Press  New York  
2532     Williams  New York  
2533      Rattray  New York  
2534     Mulligan  New York  
2535      Haskell  New York  

[2536 rows x 6 columns]

In [98]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2536 entries, 0 to 2535
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   ID          2536 non-null   object
 1   uuid        2536 non-null   object
 2   Address     2536 non-null   object
 3   First_name  2536 non-null   object
 4   Last_name   2536 non-null   object
 5   City        2528 non-null   object
dtypes: object(6)
memory usage: 119.0+ KB


다른 열에는 누락데이터가 없으나
City 열에는 2528개의 데이터가 존재하는 것으로 보아 누락데이터가 있음을 확인함.

In [124]:
8/2536

0.0031545741324921135

In [99]:
df['City'].value_counts(dropna=False)

New York            1855
Brooklyn             190
Long Island City     137
Bronx                 91
Staten Island         76
Flushing              49
Kew Gardens           47
Jamaica               30
East Elmhurst         22
None                   8
Albany                 8
Woodside               7
Corona                 3
Forest Hills           3
Kingston               2
Valhalla               1
Bayside                1
Washington             1
Astoria                1
Hauppauge              1
White Plains           1
Fresh Meadows          1
College Point          1
Name: City, dtype: int64

In [100]:
# 누락데이터 개수 확인 
df.isnull().sum(axis=0)

ID            0
uuid          0
Address       0
First_name    0
Last_name     0
City          8
dtype: int64

In [101]:
missing_df = df.isnull()
for col in missing_df.columns:
  missing_count = missing_df[col].value_counts()
  #print(missing_count)
  try:
    print(col,': ', missing_count[True]) # true가 있는것의 개수를 출력 
  except:
    print(col,": ",0)

ID :  0
uuid :  0
Address :  0
First_name :  0
Last_name :  0
City :  8


In [102]:
cnt=0
for value in df['City']:
  if value is None:
    cnt +=1

print(cnt,'개의 None 값이 있습니다.')

8 개의 None 값이 있습니다.


누락데이터의 개수가 8개 뿐이므로 city의 값을 대체할 자료가 없기 때문에 데이터 삭제로 처리

In [103]:
df_cleaned = df.dropna(subset=['City'], how='any')
len(df_cleaned)

2528

In [104]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2528 entries, 0 to 2535
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   ID          2528 non-null   object
 1   uuid        2528 non-null   object
 2   Address     2528 non-null   object
 3   First_name  2528 non-null   object
 4   Last_name   2528 non-null   object
 5   City        2528 non-null   object
dtypes: object(6)
memory usage: 138.2+ KB


---

In [105]:
df_cleaned['First_name'].value_counts(dropna=False)

Vacant       103
John          43
Michael       42
David         36
Joseph        34
            ... 
Emma           1
Alton          1
Lance          1
Phillippa      1
Darryl         1
Name: First_name, Length: 1113, dtype: int64

In [106]:
df_fvacant = df_cleaned.query("First_name == 'Vacant'")
df_fvacant

ID                                  uuid  \
6     row-3ape-4rbz-tzgp  00000000-0000-0000-CC09-20E958F99BD7   
85    row-h5nk.wrpr.2vhd  00000000-0000-0000-4A6F-BDB45D7B45D0   
86    row-9t35.wcae.w4qn  00000000-0000-0000-0ED3-645017837221   
87    row-gact_8fx5-vbi5  00000000-0000-0000-94A6-0BE2924DEBFF   
105   row-pj3g.3cx3-xbgz  00000000-0000-0000-2F8D-60A0411D52A8   
...                  ...                                   ...   
2373  row-3mav~7rkp_rez3  00000000-0000-0000-A56F-D52F19F85A41   
2414  row-rmjg.as2q~fydz  00000000-0000-0000-1602-95E947681742   
2429  row-xef5_hjvw.jvmm  00000000-0000-0000-B956-DA5A78CD95A9   
2457  row-4psz.v8cz~ik4s  00000000-0000-0000-E410-C5535A94F099   
2498  row-b9jq-s7w9-z226  00000000-0000-0000-613E-0C53F7595E8B   

                                                Address First_name Last_name  \
6     https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
85    https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
86    https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
87    https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
105   https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
...                                                 ...        ...       ...   
2373  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2414  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2429  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2457  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2498  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   

          City  
6     New York  
85       Bronx  
86       Bronx  
87       Bronx  
105   Brooklyn  
...        ...  
2373  New York  
2414  New York  
2429  New York  
2457  New York  
2498  New York  

[103 rows x 6 columns]

In [107]:
cnt=0
for value in df_cleaned['First_name']:
  if value=='Vacant':
    cnt +=1

print(cnt,'개의 Vacant(빈) 값이 있습니다.')

103 개의 Vacant(빈) 값이 있습니다.


In [109]:
df_cleaned['Last_name'].value_counts(dropna=False)

Vacant       103
Williams      14
Rivera        13
Brown         12
Rodriguez     12
            ... 
Naiburg        1
Duckfield      1
Brave          1
Sobrino        1
Haskell        1
Name: Last_name, Length: 1892, dtype: int64

In [110]:
df_lvacant = df_cleaned.query("Last_name == 'Vacant'")
df_lvacant

ID                                  uuid  \
6     row-3ape-4rbz-tzgp  00000000-0000-0000-CC09-20E958F99BD7   
85    row-h5nk.wrpr.2vhd  00000000-0000-0000-4A6F-BDB45D7B45D0   
86    row-9t35.wcae.w4qn  00000000-0000-0000-0ED3-645017837221   
87    row-gact_8fx5-vbi5  00000000-0000-0000-94A6-0BE2924DEBFF   
105   row-pj3g.3cx3-xbgz  00000000-0000-0000-2F8D-60A0411D52A8   
...                  ...                                   ...   
2373  row-3mav~7rkp_rez3  00000000-0000-0000-A56F-D52F19F85A41   
2414  row-rmjg.as2q~fydz  00000000-0000-0000-1602-95E947681742   
2429  row-xef5_hjvw.jvmm  00000000-0000-0000-B956-DA5A78CD95A9   
2457  row-4psz.v8cz~ik4s  00000000-0000-0000-E410-C5535A94F099   
2498  row-b9jq-s7w9-z226  00000000-0000-0000-613E-0C53F7595E8B   

                                                Address First_name Last_name  \
6     https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
85    https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
86    https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
87    https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
105   https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
...                                                 ...        ...       ...   
2373  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2414  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2429  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2457  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   
2498  https://data.cityofnewyork.us/resource/mdcw-n6...     Vacant    Vacant   

          City  
6     New York  
85       Bronx  
86       Bronx  
87       Bronx  
105   Brooklyn  
...        ...  
2373  New York  
2414  New York  
2429  New York  
2457  New York  
2498  New York  

[103 rows x 6 columns]

In [111]:
cnt=0
for value in df_cleaned['Last_name']:
  if value=='Vacant':
    cnt +=1

print(cnt,'개의 Vacant(빈) 값이 있습니다.')

103 개의 Vacant(빈) 값이 있습니다.


In [112]:
index_identical = df_lvacant.index.identical(df_fvacant.index)
index_identical

True

df_fvacant와 df_lvacant 의 인덱스가 일치하는 것으로 보아
Vacant의 이름이 들어있는 인덱스의 값과 순서가 동일하다는 것을 알 수 있음. 

In [113]:
df_filtered = df_cleaned[df_cleaned['First_name'] != 'Vacant']
len(df_filtered)

2425

In [114]:
df_filtered

ID                                  uuid  \
0     row-ueyx_pbr6.ht3u  00000000-0000-0000-7235-BFFC8FE5D830   
1     row-5t8b_346w~6fsq  00000000-0000-0000-33C9-8FD611F3B1E5   
2     row-y52z~x7gt_5zsp  00000000-0000-0000-0A9D-CB83FB2C9CFF   
3     row-e897~azvy.cjkb  00000000-0000-0000-4E87-A63E672BFDA8   
4     row-ie6x_vcbs_4rbs  00000000-0000-0000-F1A9-90BD9EB88D53   
...                  ...                                   ...   
2531  row-w2ye~6ae8.ip8q  00000000-0000-0000-E302-06F2992D254C   
2532  row-nene.dygq.e72g  00000000-0000-0000-EE13-E543BEBD2BDD   
2533  row-8d7b-bf86_jrq9  00000000-0000-0000-0B1A-8CCF3CF56D98   
2534  row-nzmw_4y5g-htgt  00000000-0000-0000-B8D2-70472FD59060   
2535  row-q2yj~844r_3byi  00000000-0000-0000-724C-63423CED4A0E   

                                                Address First_name  \
0     https://data.cityofnewyork.us/resource/mdcw-n6...    Annette   
1     https://data.cityofnewyork.us/resource/mdcw-n6...      Keith   
2     https://data.cityofnewyork.us/resource/mdcw-n6...      Marek   
3     https://data.cityofnewyork.us/resource/mdcw-n6...    Marlene   
4     https://data.cityofnewyork.us/resource/mdcw-n6...    Melissa   
...                                                 ...        ...   
2531  https://data.cityofnewyork.us/resource/mdcw-n6...   Caroline   
2532  https://data.cityofnewyork.us/resource/mdcw-n6...     Denice   
2533  https://data.cityofnewyork.us/resource/mdcw-n6...     Darryl   
2534  https://data.cityofnewyork.us/resource/mdcw-n6...    Valerie   
2535  https://data.cityofnewyork.us/resource/mdcw-n6...      Susan   

        Last_name      City  
0         Charles  New York  
1            Snow  New York  
2     Tyszkiewicz  New York  
3     Markoe-Boyd  New York  
4          Chacko  New York  
...           ...       ...  
2531        Press  New York  
2532     Williams  New York  
2533      Rattray  New York  
2534     Mulligan  New York  
2535      Haskell  New York  

[2425 rows x 6 columns]

이름이 'Vacant'를 제외한 실질적으로 이용 할 수 있는 data의 rows의 개수는 2425개

---

In [118]:
# 'First_name' 이름 count 내림차순 
df_filtered['First_name'].value_counts(dropna=False)

John         43
Michael      42
David        36
Joseph       34
Robert       26
             ..
Emma          1
Alton         1
Lance         1
Phillippa     1
Darryl        1
Name: First_name, Length: 1112, dtype: int64

In [122]:
# CIty별 'First_name' 최빈값
most_common_first_names = df_filtered.groupby('City')['First_name'].apply(lambda x: x.value_counts().idxmax()).reset_index()

most_common_first_names

City   First_name
0             Albany  Christopher
1            Astoria         Carl
2            Bayside        Cesar
3              Bronx        Lloyd
4           Brooklyn         John
5      College Point        Peter
6             Corona     Terrance
7      East Elmhurst       Ronald
8           Flushing       Joseph
9       Forest Hills       Amanda
10     Fresh Meadows        Kevin
11         Hauppauge    Catherine
12           Jamaica       Marsha
13       Kew Gardens         John
14          Kingston          Joe
15  Long Island City     Michelle
16          New York      Michael
17     Staten Island       Daniel
18          Valhalla         Paul
19        Washington      Crystal
20      White Plains         Gary
21          Woodside      Anthony